plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,324 @@
+}
+
+// SampleConfig will populate the sample configuration portion of the plugin's configuration
+func (xio *XtremIO) SampleConfig() string {
+	return sampleConfig
+}
+
+func (xio *XtremIO) Init() error {
+	if xio.Username == "" {
+		return fmt.Errorf("Username cannot be empty")
+	}
+	if xio.Password == "" {
+		return fmt.Errorf("Password cannot be empty")
+	}
+	if xio.URL == "" {
+		return fmt.Errorf("URL cannot be empty")
+	}
+
+	return nil
+}
+
+func (xio *XtremIO) Gather(acc telegraf.Accumulator) error {
+	tlsCfg, err := xio.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = tlsCfg
+	err = xio.Authenticate()
+	if err != nil {
+		return err
+	}
+	if xio.Cookie == nil {
+		return fmt.Errorf("no authentication cookie set")
+	}
+
+	availableCollectors := []string{"bbus", "clusters", "ssds", "volumes", "xms"}
+	var wg sync.WaitGroup
+	for _, collector := range xio.Collectors {
+		wg.Add(1)
+		go func(collector string) {
+			defer wg.Done()
+			if !contains(availableCollectors, collector) {
+				acc.AddError(fmt.Errorf("Specified Collector Isn't Supported: " + collector))
+				return
+			}
+
+			resp, err := xio.Call(collector)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+			var arr []gjson.Result
+			if collector == "xms" {
+				arr = gjson.Get(resp, "xmss").Array()
+			} else {
+				arr = gjson.Get(resp, collector).Array()
+			}
+
+			for _, item := range arr {
+				itemSplit := strings.Split(gjson.Get(item.Raw, "href").Str, "/")
+				url := ""
+				if len(itemSplit) > 0 {
+					url = collector + "/" + itemSplit[len(itemSplit)-1]
+				} else {
+					continue
+				}
+
+				if collector == "bbus" {
+					wg.Add(1)
+					go xio.GatherBBUs(&wg, url, acc)
+				}
+				if collector == "clusters" {
+					wg.Add(1)
+					go xio.GatherClusters(&wg, url, acc)
+				}
+				if collector == "ssds" {
+					wg.Add(1)
+					go xio.GatherSSDs(&wg, url, acc)
+				}
+				if collector == "volumes" {
+					wg.Add(1)
+					go xio.GatherVolumes(&wg, url, acc)
+				}
+				if collector == "xms" {
+					wg.Add(1)
+					go xio.GatherXMS(&wg, url, acc)
+				}
+			}
+		}(collector)
+	}
+
+	wg.Wait()
+
+	xio.ResetCookie()
+
+	return nil
+}
+
+func (xio *XtremIO) GatherBBUs(wg *sync.WaitGroup, url string, acc telegraf.Accumulator) {
+	defer wg.Done()
+	resp, _ := xio.Call(url)
+	tags := map[string]string{
+		"serial_number": gjson.Get(resp, "content.serial-number").Str,
+		"guid":          gjson.Get(resp, "content.guid").Str,
+		"power_feed":    gjson.Get(resp, "content.power-feed").Str,
+		"name":          gjson.Get(resp, "content.name").Str,
+		"model_name":    gjson.Get(resp, "content.model-name").Str,
+	}
+	fields := make(map[string]interface{})
+	fields["bbus_power"], _ = strconv.Atoi(gjson.Get(resp, "content.power").Raw)
+	fields["bbus_average_daily_temp"], _ = strconv.Atoi(gjson.Get(resp, "content.avg-daily-temp").Raw)
```

Comment:
Just suppressing errors won't make them go away! Especially those conversion errors point to real problems that otherwise go unnoticed. If you for example are parsing a string field here you will get an error (which you suppress) and a return value of zero. The user then only sees `bbus_power = 0` and cannot decide if this is the actual value or an erroneous parsing result.

Reply:
Take a look at what I did here with the errors and see if it's alright with you. I tried to find a way to do it without putting an if statement after every assignment, since that seems like something that should be in a loop at that point, but I didn't find much. I'm not a Go developer, so if you know of a better way to do this please let me know.